#Quiz
> 공공자전거 대여 이력 정보 데이터 중 이용거리에 대한 결측치를 처리하시오

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [24]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Lec4_BigData/1. 데이터분석을 위한 기초 통계학/1.데이터 클렌징/data/bicycle.csv', encoding='CP949')
df

자전거번호              대여일시  대여소번호  ... 반납거치대  이용시간    이용거리
0    SPB-23220   2019-11-01 8:48    646  ...   7.0  12.0  1100.0
1    SPB-16216   2019-11-04 8:38    646  ...   2.0   7.0  1420.0
2    SPB-21097   2019-11-04 8:46    646  ...   7.0  10.0     NaN
3    SPB-22292   2019-11-05 8:34    646  ...   NaN  10.0  1380.0
4    SPB-07935  2019-11-05 12:29    512  ...   7.0  10.0  1650.0
..         ...               ...    ...  ...   ...   ...     ...
415  SPB-12116   2019-11-15 4:22    112  ...   1.0  11.0  2200.0
416  SPB-17768   2019-11-15 9:31    103  ...   NaN   5.0     NaN
417  SPB-18841  2019-11-15 20:40    113  ...   1.0  12.0  2280.0
418  SPB-24467  2019-11-15 21:55    152  ...   1.0  10.0  1180.0
419  SPB-17570  2019-11-15 22:16    391  ...   5.0  48.0  8100.0

[420 rows x 11 columns]

In [6]:
df[df['대여일시'].isna()]

Empty DataFrame
Columns: [자전거번호, 대여일시, 대여소번호, 대여소명, 대여거치대, 반납일시, 반납대여소번호, 반납대여소명, 반납거치대, 이용시간, 이용거리]
Index: []

In [7]:
df[df['이용거리'].isna()]

자전거번호              대여일시  대여소번호  ... 반납거치대  이용시간 이용거리
2    SPB-21097   2019-11-04 8:46    646  ...   7.0  10.0  NaN
12   SPB-19057   2019-11-13 8:55    646  ...   7.0  11.0  NaN
23   SPB-19256  2019-11-26 10:20      3  ...   NaN   3.0  NaN
37   SPB-14246   2019-11-01 2:21    120  ...   2.0  23.0  NaN
70   SPB-08265  2019-11-02 14:24    104  ...   1.0  28.0  NaN
94   SPB-11003  2019-11-02 23:45    102  ...   2.0   6.0  NaN
128  SPB-03349  2019-11-03 23:03    124  ...   3.0  19.0  NaN
179  SPB-14571   2019-11-05 9:09    438  ...   5.0  17.0  NaN
252  SPB-07514  2019-11-07 18:59    225  ...   5.0  29.0  NaN
264  SPB-20868   2019-11-08 9:46    104  ...   4.0   7.0  NaN
279  SPB-14276  2019-11-08 19:07    410  ...   4.0  25.0  NaN
295  SPB-16380   2019-11-09 0:35    419  ...   1.0  16.0  NaN
315  SPB-24164  2019-11-10 10:04    182  ...   2.0  35.0  NaN
372  SPB-06102  2019-11-12 18:55    151  ...   2.0   6.0  NaN
409  SPB-10593  2019-11-14 23:29    106  ...   4.0  10.0  NaN
416  SPB-17768   2019-11-15 9:31    103  ...   NaN   5.0  NaN

[16 rows x 11 columns]

In [8]:
distance_per_usetime = df.groupby('이용시간').mean()['이용거리']
distance_per_usetime

이용시간
2.0         20.000000
3.0        660.000000
4.0        804.000000
5.0        849.062500
6.0       1002.647059
             ...     
105.0     1050.000000
106.0     3110.000000
114.0     2970.000000
121.0    16985.000000
122.0     3440.000000
Name: 이용거리, Length: 69, dtype: float64

In [9]:
distance_per_usetime = distance_per_usetime.loc[df[df['이용거리'].isna()]['이용시간']]
distance_per_usetime

이용시간
10.0    1393.333333
11.0    1538.235294
3.0      660.000000
23.0    4360.000000
28.0    4545.555556
6.0     1002.647059
19.0    2806.666667
17.0    2232.222222
29.0    5500.000000
7.0     1084.117647
25.0    3634.166667
16.0    2257.777778
35.0    4015.000000
6.0     1002.647059
10.0    1393.333333
5.0      849.062500
Name: 이용거리, dtype: float64

In [10]:
df_test = df
index = list((distance_per_usetime.index))
val = distance_per_usetime.values

print(index)
print(val)

[10.0, 11.0, 3.0, 23.0, 28.0, 6.0, 19.0, 17.0, 29.0, 7.0, 25.0, 16.0, 35.0, 6.0, 10.0, 5.0]
[1393.33333333 1538.23529412  660.         4360.         4545.55555556
 1002.64705882 2806.66666667 2232.22222222 5500.         1084.11764706
 3634.16666667 2257.77777778 4015.         1002.64705882 1393.33333333
  849.0625    ]


In [11]:
df_test[df_test['이용거리'].isna()]

자전거번호              대여일시  대여소번호  ... 반납거치대  이용시간 이용거리
2    SPB-21097   2019-11-04 8:46    646  ...   7.0  10.0  NaN
12   SPB-19057   2019-11-13 8:55    646  ...   7.0  11.0  NaN
23   SPB-19256  2019-11-26 10:20      3  ...   NaN   3.0  NaN
37   SPB-14246   2019-11-01 2:21    120  ...   2.0  23.0  NaN
70   SPB-08265  2019-11-02 14:24    104  ...   1.0  28.0  NaN
94   SPB-11003  2019-11-02 23:45    102  ...   2.0   6.0  NaN
128  SPB-03349  2019-11-03 23:03    124  ...   3.0  19.0  NaN
179  SPB-14571   2019-11-05 9:09    438  ...   5.0  17.0  NaN
252  SPB-07514  2019-11-07 18:59    225  ...   5.0  29.0  NaN
264  SPB-20868   2019-11-08 9:46    104  ...   4.0   7.0  NaN
279  SPB-14276  2019-11-08 19:07    410  ...   4.0  25.0  NaN
295  SPB-16380   2019-11-09 0:35    419  ...   1.0  16.0  NaN
315  SPB-24164  2019-11-10 10:04    182  ...   2.0  35.0  NaN
372  SPB-06102  2019-11-12 18:55    151  ...   2.0   6.0  NaN
409  SPB-10593  2019-11-14 23:29    106  ...   4.0  10.0  NaN
416  SPB-17768   2019-11-15 9:31    103  ...   NaN   5.0  NaN

[16 rows x 11 columns]

In [12]:
for i in range(len(index)):
  print(index[i], val[i])
  df_test[(df_test['이용거리'].isnull()) & (df_test['이용시간'] == index[i])] = df_test[(df_test['이용거리'].isnull()) & (df_test['이용시간'] == index[i])].replace({'이용거리':np.NaN}, {'이용거리':val[i]})

10.0 1393.3333333333333
11.0 1538.235294117647
3.0 660.0
23.0 4360.0
28.0 4545.555555555556
6.0 1002.6470588235294
19.0 2806.6666666666665
17.0 2232.222222222222
29.0 5500.0
7.0 1084.1176470588234
25.0 3634.1666666666665
16.0 2257.777777777778
35.0 4015.0
6.0 1002.6470588235294
10.0 1393.3333333333333
5.0 849.0625


In [13]:
df_test

자전거번호              대여일시  대여소번호  ... 반납거치대  이용시간         이용거리
0    SPB-23220   2019-11-01 8:48    646  ...   7.0  12.0  1100.000000
1    SPB-16216   2019-11-04 8:38    646  ...   2.0   7.0  1420.000000
2    SPB-21097   2019-11-04 8:46    646  ...   7.0  10.0  1393.333333
3    SPB-22292   2019-11-05 8:34    646  ...   NaN  10.0  1380.000000
4    SPB-07935  2019-11-05 12:29    512  ...   7.0  10.0  1650.000000
..         ...               ...    ...  ...   ...   ...          ...
415  SPB-12116   2019-11-15 4:22    112  ...   1.0  11.0  2200.000000
416  SPB-17768   2019-11-15 9:31    103  ...   NaN   5.0   849.062500
417  SPB-18841  2019-11-15 20:40    113  ...   1.0  12.0  2280.000000
418  SPB-24467  2019-11-15 21:55    152  ...   1.0  10.0  1180.000000
419  SPB-17570  2019-11-15 22:16    391  ...   5.0  48.0  8100.000000

[420 rows x 11 columns]

In [14]:
df_test[df_test['이용거리'].isnull()]

Empty DataFrame
Columns: [자전거번호, 대여일시, 대여소번호, 대여소명, 대여거치대, 반납일시, 반납대여소번호, 반납대여소명, 반납거치대, 이용시간, 이용거리]
Index: []

#강사님코드

>이용 거리는 출발지에서 도착지까지 이동한 경로에 대한 거리로 최단 경로로 이동할 수도 있고, 경유를 하여 이동할 수 있다. 이에따라 이용거리는 달라진다.

> 이용거리의 평균을 구하여 대표값으로 사용하기에는 출발지와 도착지가 모두 다름으로 대표값으로 사용할 수 없게 된다.

> 따라서 출발지가 같은 지역의 이용거리 평균을 이용하여 결측치를 대체할 것이다. (단, 같은 출발지의 도착지는 같다고 가정한다.)

> 첫 줄의 대여소 번호는 출발지를 뜻하는 것으로 출발지가 646인 모든 결과를 출력하는 코드이다.

> 예를들어 출발지가 646인값을 확인하면 반납대여소번호가 3번으로 도착지가 같고 이용시간은 최소 5분에서 최대30분이 소요되는 것을 볼 수 있으며, 이용거리는 1060미터에서 최대 1420미터 이동하는 것을 알 수 있다.

> 따라서 같은 출발지이면 최대, 최소 이동거리의 차가 크지 않음으로 이의 평균 거리를 결측치도 대체하는것은 타당하다고 판단할 수 있따

In [15]:
# df2 = df
# df2.replace(np.nan, 'test')

df2 = df
df2.replace({'이용거리':np.NaN}, {'이용거리':'test'})

자전거번호              대여일시  대여소번호  ... 반납거치대  이용시간         이용거리
0    SPB-23220   2019-11-01 8:48    646  ...   7.0  12.0  1100.000000
1    SPB-16216   2019-11-04 8:38    646  ...   2.0   7.0  1420.000000
2    SPB-21097   2019-11-04 8:46    646  ...   7.0  10.0  1393.333333
3    SPB-22292   2019-11-05 8:34    646  ...   NaN  10.0  1380.000000
4    SPB-07935  2019-11-05 12:29    512  ...   7.0  10.0  1650.000000
..         ...               ...    ...  ...   ...   ...          ...
415  SPB-12116   2019-11-15 4:22    112  ...   1.0  11.0  2200.000000
416  SPB-17768   2019-11-15 9:31    103  ...   NaN   5.0   849.062500
417  SPB-18841  2019-11-15 20:40    113  ...   1.0  12.0  2280.000000
418  SPB-24467  2019-11-15 21:55    152  ...   1.0  10.0  1180.000000
419  SPB-17570  2019-11-15 22:16    391  ...   5.0  48.0  8100.000000

[420 rows x 11 columns]

> 부분 교체를 위해 replace함수에 대한 사용법을 알아보자
* replace(찾을 데이터, 변경할 데이터)

>NaN값을 찾을경우 numpy에서 지원하는 nan을 이용하면 된다.
특정 feature를 지정하려면 dict형식을 이용하면 된다.

In [16]:
df2 = df

#df2
# 대여소가 646번인 값을 추출한다.

# 대여소번호가 646번인 이용거리의 평균
mean646 = df2[df2['대여소번호']==646]['이용거리'].mean()
print(mean646)
#대여소번호가 646번인 데이터중 결측치를 평균으로 대체
df2[df['대여소번호']==646].replace({'이용거리' : np.nan}, {'이용거리':mean646})

#대여소 번호가 646번인 dataframe에 대여소 번호가 646번인 결특치를 대체한 결과를 저장
#mean646 = df2[]


1276.2651727357609


자전거번호             대여일시  대여소번호  ... 반납거치대  이용시간         이용거리
0   SPB-23220  2019-11-01 8:48    646  ...   7.0  12.0  1100.000000
1   SPB-16216  2019-11-04 8:38    646  ...   2.0   7.0  1420.000000
2   SPB-21097  2019-11-04 8:46    646  ...   7.0  10.0  1393.333333
3   SPB-22292  2019-11-05 8:34    646  ...   NaN  10.0  1380.000000
5   SPB-18401  2019-11-06 8:52    646  ...   7.0  11.0  1350.000000
6   SPB-18350  2019-11-06 8:35    646  ...   2.0   9.0  1390.000000
8   SPB-20220  2019-11-11 8:35    646  ...   2.0   6.0  1160.000000
9   SPB-20102  2019-11-11 9:45    646  ...   7.0  30.0  1060.000000
10  SPB-10000  2019-11-12 8:35    646  ...   2.0   8.0  1300.000000
11  SPB-20145  2019-11-13 8:30    646  ...   2.0   NaN  1060.000000
12  SPB-19057  2019-11-13 8:55    646  ...   7.0  11.0  1538.235294
13  SPB-19645  2019-11-14 9:44    646  ...   7.0  16.0  1140.000000
14  SPB-21851  2019-11-14 9:33    646  ...   2.0  28.0  1220.000000
15  SPB-13973  2019-11-15 8:42    646  ...   2.0   6.0  1370.000000
16  SPB-22311  2019-11-18 8:32    646  ...   2.0   NaN  1350.000000
17  SPB-02270  2019-11-18 8:49    646  ...   7.0  11.0  1350.000000
18  SPB-23659  2019-11-21 8:36    646  ...   2.0   6.0  1130.000000
19  SPB-23497  2019-11-21 8:50    646  ...   7.0  13.0  1330.000000
21  SPB-21810  2019-11-25 8:41    646  ...   2.0   7.0  1310.000000
22  SPB-05857  2019-11-26 8:40    646  ...   2.0   NaN  1380.000000
24  SPB-20076  2019-11-28 8:37    646  ...   2.0   5.0  1070.000000

[21 rows x 11 columns]

## 결측치 처리 순서
1. 결측열 찾기
2. 대여소 번호 얻기
3. 대여소 번호와 column명을 전달하여 변경하기

#3. Quiz
> 함수화 처리방법
  1. 하나의 데이터를 대상으로 정상적으로 동작되는지 확인
  2. 처리된 코드를 def 함수명(): dkfo sjgdjwnsek.
  3. 고정값과 가변값 선별
  4. 가변값은 def 함수명(가변값): 형식으로 변경
  5. 반환 값은 return으로 처리

def test():
  mean646 = df2[df2['대여소번호']== 646]['이용거리'].mean()
  df2[df2['대여소번호']==646] = df2[df['대여소번호']==646].replace({'이용거리' : np.nan}, {'이용거리':mean646})

In [17]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [18]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Lec4_BigData/1. 데이터분석을 위한 기초 통계학/1.데이터 클렌징/data/bicycle.csv', encoding='CP949')
df2 = df

In [19]:
df2[df2['이용거리'].isnull()]

자전거번호              대여일시  대여소번호  ... 반납거치대  이용시간 이용거리
2    SPB-21097   2019-11-04 8:46    646  ...   7.0  10.0  NaN
12   SPB-19057   2019-11-13 8:55    646  ...   7.0  11.0  NaN
23   SPB-19256  2019-11-26 10:20      3  ...   NaN   3.0  NaN
37   SPB-14246   2019-11-01 2:21    120  ...   2.0  23.0  NaN
70   SPB-08265  2019-11-02 14:24    104  ...   1.0  28.0  NaN
94   SPB-11003  2019-11-02 23:45    102  ...   2.0   6.0  NaN
128  SPB-03349  2019-11-03 23:03    124  ...   3.0  19.0  NaN
179  SPB-14571   2019-11-05 9:09    438  ...   5.0  17.0  NaN
252  SPB-07514  2019-11-07 18:59    225  ...   5.0  29.0  NaN
264  SPB-20868   2019-11-08 9:46    104  ...   4.0   7.0  NaN
279  SPB-14276  2019-11-08 19:07    410  ...   4.0  25.0  NaN
295  SPB-16380   2019-11-09 0:35    419  ...   1.0  16.0  NaN
315  SPB-24164  2019-11-10 10:04    182  ...   2.0  35.0  NaN
372  SPB-06102  2019-11-12 18:55    151  ...   2.0   6.0  NaN
409  SPB-10593  2019-11-14 23:29    106  ...   4.0  10.0  NaN
416  SPB-17768   2019-11-15 9:31    103  ...   NaN   5.0  NaN

[16 rows x 11 columns]

In [20]:
def FillNoneVal(placenum):
  distance_mean = df2[df2['대여소번호']==placenum]['이용거리'].mean()
  df2[df2['대여소번호']==placenum] = df2[df['대여소번호']==placenum].replace({'이용거리' : np.nan}, {'이용거리':distance_mean})
  print("placenum: ",placenum )
  print("distance_mean: ",distance_mean )

In [21]:
value_list = list(df2[df2['이용거리'].isnull()]['대여소번호'].values)

for item in value_list:
  FillNoneVal(item)

placenum:  646
distance_mean:  1256.3157894736842
placenum:  646
distance_mean:  1256.3157894736842
placenum:  3
distance_mean:  0.0
placenum:  120
distance_mean:  nan
placenum:  104
distance_mean:  1085.3333333333333
placenum:  102
distance_mean:  1376.4705882352941
placenum:  124
distance_mean:  2790.0
placenum:  438
distance_mean:  3900.0
placenum:  225
distance_mean:  nan
placenum:  104
distance_mean:  1085.3333333333333
placenum:  410
distance_mean:  5020.0
placenum:  419
distance_mean:  2930.0
placenum:  182
distance_mean:  4960.0
placenum:  151
distance_mean:  1344.0
placenum:  106
distance_mean:  806.5625
placenum:  103
distance_mean:  1425.5172413793102


In [22]:
df2[df2['이용거리'].isnull()]

자전거번호              대여일시  대여소번호  ... 반납거치대  이용시간 이용거리
37   SPB-14246   2019-11-01 2:21    120  ...   2.0  23.0  NaN
252  SPB-07514  2019-11-07 18:59    225  ...   5.0  29.0  NaN

[2 rows x 11 columns]

#강사님 코드

> 함수화 방법
1. 외부에서 전달되는 모든 값은 인자로 처리한다.
2. 내부에서 중복되는 값은 별도의 변수로 처리한다.
3. 결과는 return을 이용하여 반환한다.

##1. 결측렬 찾기

In [25]:
#결측치인 값이 존재하면 true로 반환
df.isnull()

자전거번호   대여일시  대여소번호   대여소명  대여거치대  ...  반납대여소번호  반납대여소명  반납거치대   이용시간   이용거리
0    False  False  False  False  False  ...    False   False  False  False  False
1    False  False  False  False  False  ...    False   False  False  False  False
2    False  False  False  False  False  ...    False   False  False  False   True
3    False  False  False  False  False  ...    False   False   True  False  False
4    False  False  False  False  False  ...    False   False  False  False  False
..     ...    ...    ...    ...    ...  ...      ...     ...    ...    ...    ...
415  False  False  False  False  False  ...    False   False  False  False  False
416  False  False  False  False  False  ...    False   False   True  False   True
417  False  False  False  False  False  ...    False   False  False  False  False
418  False  False  False  False  False  ...    False   False  False  False  False
419  False  False  False  False  False  ...    False   False  False  False  False

[420 rows x 11 columns]

In [26]:
#feature별 결측치 집계
df.isnull().sum()

자전거번호       0
대여일시        0
대여소번호       0
대여소명        0
대여거치대      24
반납일시        0
반납대여소번호     0
반납대여소명      0
반납거치대      11
이용시간       15
이용거리       16
dtype: int64

In [38]:
def null_feature_indexs():
  features = df.isnull().sum()

  #결측치가 존재하는 feature의 index정보 추출
  null_features = features[features>0]
  return null_features.index.to_list() #리스트로 반환`

In [36]:
print(null_feature_indexs())

['대여거치대', '반납거치대', '이용시간', '이용거리']


In [41]:
df['이용거리']
df[df['이용거리']==np.nan]

Empty DataFrame
Columns: [자전거번호, 대여일시, 대여소번호, 대여소명, 대여거치대, 반납일시, 반납대여소번호, 반납대여소명, 반납거치대, 이용시간, 이용거리]
Index: []

In [42]:
#대여소번호 중 중복된 데이터 제거
df['대여소번호'].drop_duplicates()

0       646
4       512
7         3
20      529
25      419
       ... 
347     403
348     346
350     245
352     623
398    1160
Name: 대여소번호, Length: 76, dtype: int64

In [45]:
def getNolist(data_frame):
  no_series = df['대여소번호'].drop_duplicates()
  return list(no_series.values)

nolst = getNolist(df)
nolst

[646,
 512,
 3,
 529,
 419,
 911,
 108,
 199,
 102,
 101,
 120,
 106,
 421,
 956,
 104,
 103,
 152,
 125,
 113,
 148,
 226,
 128,
 267,
 115,
 826,
 391,
 105,
 137,
 154,
 409,
 116,
 151,
 272,
 175,
 182,
 3102,
 112,
 129,
 228,
 139,
 2103,
 183,
 114,
 440,
 184,
 196,
 124,
 119,
 264,
 438,
 203,
 432,
 121,
 109,
 306,
 111,
 181,
 427,
 247,
 224,
 410,
 413,
 1925,
 141,
 205,
 408,
 225,
 186,
 284,
 134,
 1998,
 403,
 346,
 245,
 623,
 1160]

In [ ]:
def ReplaceNaN(data_frame, rentNo):
  bool_rentNo = data_frame['대여소번호']==rentNo
  mean_distance = round(data_frame[bool_rentNo]['이동거리'].mean(). 0)
  data_frame[bool_rentNo] = data_frame[bool_rentNo].replace({})

  return data_frame

In [ ]:
# for row in lst:
  #for no in nolst:
    #for print(row, " : ", no)

for no in nolst:
  df = ReplaceNaN(df, no)

df